### PR TITLE
[ADD] add a new module:stock_pack_wizard which allows you packing different stock moves in one picking together and quantity you want.

### PR DESCRIPTION
**Pull request by duanyp1991**
```
    add a new module:stock_pack_wizard which allows you packing different stock moves in one picking together and quantity you want.
 Changes to be committed:
    new file:   stock_pack_wizard/__init__.py
    new file:   stock_pack_wizard/__openerp__.py
    new file:   stock_pack_wizard/data/product.ul.csv
    new file:   stock_pack_wizard/pack_data.xml
    new file:   stock_pack_wizard/product_ul.py
    new file:   stock_pack_wizard/product_ul_view.xml
    new file:   stock_pack_wizard/report/__init__.py
    new file:   stock_pack_wizard/report/stock_tracking_list.py
    new file:   stock_pack_wizard/report/stock_trackling_list.mako
    new file:   stock_pack_wizard/report/stock_trackling_list.rml
    new file:   stock_pack_wizard/report/stock_trackling_list.sxw
    new file:   stock_pack_wizard/security/ir.model.access.csv
    new file:   stock_pack_wizard/stock_picking_view.xml
    new file:   stock_pack_wizard/stock_tracking.py
    new file:   stock_pack_wizard/stock_tracking_report.xml
    new file:   stock_pack_wizard/stock_tracking_view.xml
    new file:   stock_pack_wizard/test/test_pack_wizard.yml
    new file:   stock_pack_wizard/wizard/__init__.py
    new file:   stock_pack_wizard/wizard/wizard_picking_tracking.py
    new file:   stock_pack_wizard/wizard/wizard_picking_tracking_view.xml
```

### CODE REVIEW

**Comment by gurneyalex** [comment]
CLA checked OK

File: stock_pack_wizard/__openerp__.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+============
+
+This module depends on **sale_stock** and **report_webkit** modules.
+
+Usage
+=====
+To use this module, you need to:
+
+* The current user should belong to the group:
+* Go to Settings > Warehosue >
```
Comment (gurneyalex) [inline stock_pack_wizard/__openerp__.py:50]: s/Warehosue/Warehouse/

File: stock_pack_wizard/__openerp__.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    "Track serial number on logistic units (pallets)"
+* Go to Warehosue > delivery order
```
Comment (gurneyalex) [inline stock_pack_wizard/__openerp__.py:52]: same

File: stock_pack_wizard/data/product.ul.csv [hunk 3]
```diff
@@ -0,0 +1,32 @@
+id,name,high,width,long,type
+pack_1, SPARK-1/43-12 - SPARK新包装中箱, 23.000 , 18.000 , 27.000,pack
```
Comment (gurneyalex) [inline stock_pack_wizard/data/product.ul.csv:2]: This file needs either to be translated to english or moved to a separate module part of a chinese localization.

File: stock_pack_wizard/product_ul.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+
+class ProductUl(orm.Model):
+    _inherit = 'product.ul'
+
+    def _get_cbm(self, cr, uid, ids, fields, arg=None, context=None):
+        res = {}
+        for ul in self.browse(cr, uid, ids, context=context):
+            cbm = ul.high * ul.width * ul.long
+            cbm = cbm != 0 and cbm / 1000000
```
Comment (gurneyalex) [inline stock_pack_wizard/product_ul.py:35]: this assumes the dimensions are in centimeters. The product information about dimensions in Odoo is in meters, and I think you should stick to this.

File: stock_pack_wizard/product_ul.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+            res[ul.id] = cbm
+        return res
+
+    _columns = {
+        # re-define this field to change the size.
+        'name': fields.char(
+            'name', size=32,
+            select=True, required=True, translate=True),
+        'high': fields.float(
+            'H (cm)', digits_compute=dp.get_precision('Pack Height')),
```
Comment (gurneyalex) [inline stock_pack_wizard/product_ul.py:45]: not sure about using cm here (see above)

File: stock_pack_wizard/product_ul.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+            cbm = cbm != 0 and cbm / 1000000
+            res[ul.id] = cbm
+        return res
+
+    _columns = {
+        # re-define this field to change the size.
+        'name': fields.char(
+            'name', size=32,
+            select=True, required=True, translate=True),
+        'high': fields.float(
```
Comment (gurneyalex) [inline stock_pack_wizard/product_ul.py:44]: s/high/height/

File: stock_pack_wizard/product_ul.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+            'H (cm)', digits_compute=dp.get_precision('Pack Height')),
+        'width': fields.float(
+            'W (cm)', digits_compute=dp.get_precision('Pack Height')),
+        'long': fields.float(
```
Comment (gurneyalex) [inline stock_pack_wizard/product_ul.py:48]: s/long/length/

File: stock_pack_wizard/stock_tracking.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+        for pack in self.browse(cr, uid, ids, context=context):
+            res[pack.id] = sum(
+                [m.product_id.weight_net * m.product_qty
+                    for m in pack.move_ids])
+        return res
+
+    _columns = {
+        'ul_id': fields.many2one('product.ul', 'Pack Template'),
+        'pack_h': fields.related(
+            'ul_id', 'high', string='H (cm)',
```
Comment (gurneyalex) [inline stock_pack_wizard/stock_tracking.py:48]: I'd user meters here

File: stock_pack_wizard/stock_tracking.py [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+        res = {}
+        for pack in self.browse(cr, uid, ids, context=context):
+            res[pack.id] = sum(
+                [m.product_id.weight_net * m.product_qty
+                    for m in pack.move_ids])
+        return res
+
+    _columns = {
+        'ul_id': fields.many2one('product.ul', 'Pack Template'),
+        'pack_h': fields.related(
```
Comment (gurneyalex) [inline stock_pack_wizard/stock_tracking.py:47]: s/pack_h/pack_height/

File: stock_pack_wizard/stock_tracking.py [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+            'ul_id', 'high', string='H (cm)',
+            type='float',
+            digits_compute=dp.get_precision('Pack Height'),
+            readonly=True),
+        'pack_w': fields.related(
```
Comment (gurneyalex) [inline stock_pack_wizard/stock_tracking.py:52]: pack_width

File: stock_pack_wizard/stock_tracking.py [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+            'ul_id', 'width', string='W (cm)',
+            type='float',
+            digits_compute=dp.get_precision('Pack Height'),
+            readonly=True),
+        'pack_l': fields.related(
```
Comment (gurneyalex) [inline stock_pack_wizard/stock_tracking.py:57]: pack_length

File: stock_pack_wizard/stock_tracking.py [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+            type='float',
+            digits_compute=dp.get_precision('Pack Height'), readonly=True),
+        'pack_cbm': fields.related(
+            'ul_id', 'cbm', string='CBM', type='float',
+            digits_compute=dp.get_precision('Pack Height'), readonly=True),
+        'pack_address': fields.char('Address', size=128),
+        'pack_note': fields.char('Note', size=128),
+
+        'gross_weight': fields.float(
+            'GW (Kg)',
```
Comment (gurneyalex) [inline stock_pack_wizard/stock_tracking.py:68]: s/Kg/kg/

File: stock_pack_wizard/stock_tracking.py [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+            digits_compute=dp.get_precision('Pack Weight')),
+        'net_weight': fields.function(
+            _get_net_weight, arg=None,
+            type='float', string='NW (Kg)',
```
Comment (gurneyalex) [inline stock_pack_wizard/stock_tracking.py:72]: s/Kg/kg/

File: stock_pack_wizard/stock_tracking.py [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+            digits_compute=dp.get_precision('Pack Height'),
+            readonly=True),
+        'pack_l': fields.related(
+            'ul_id', 'long', string='L (cm)',
+            type='float',
+            digits_compute=dp.get_precision('Pack Height'), readonly=True),
+        'pack_cbm': fields.related(
+            'ul_id', 'cbm', string='CBM', type='float',
+            digits_compute=dp.get_precision('Pack Height'), readonly=True),
+        'pack_address': fields.char('Address', size=128),
```
Comment (gurneyalex) [inline stock_pack_wizard/stock_tracking.py:64]: don't put an arbitrary length on the field, and let postgresql handle this and optimize it.

File: stock_pack_wizard/stock_tracking.py [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+        'pack_note': fields.char('Note', size=128),
```
Comment (gurneyalex) [inline stock_pack_wizard/stock_tracking.py:65]: don't put an arbitrary length on the field, and let postgresql handle this and optimize it.

File: stock_pack_wizard/report/stock_tracking_list.py [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+
+    def _get_total_qty(self, tracking_id):
+        tracking = self.pool.get('stock.tracking').browse(self.cr, self.uid, tracking_id)
+        return sum([line.product_qty for line in tracking.move_ids])
+
+    def _get_measurement(self, tracking_id):
+        tracking = self.pool.get('stock.tracking').browse(self.cr, self.uid, tracking_id)
+        return r'*'.join([str(tracking.pack_h), str(tracking.pack_w), str(tracking.pack_l)] )
+
+    def _get_cbm(self, tracking_id):
```
Comment (gurneyalex) [inline stock_pack_wizard/report/stock_tracking_list.py:63]: why not use tracking.pack_cbm?

File: stock_pack_wizard/product_ul.py [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+            'L (cm)', digits_compute=dp.get_precision('Pack Height')),
+        'cbm': fields.function(
+            _get_cbm, arg=None, type='float',
+            string='CBM',
+            digits_compute=dp.get_precision('Pack Height')),
```
Comment (gurneyalex) [inline stock_pack_wizard/product_ul.py:53]: you should store this field.

File: stock_pack_wizard/report/stock_tracking_list.py [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+        customs_description = product.customs_description and ' \n\n\n' + product.customs_description or ''
+        return ''.join([default_code, customs_description, hs_code])
+
+    def _get_total_qty(self, tracking_id):
+        tracking = self.pool.get('stock.tracking').browse(self.cr, self.uid, tracking_id)
+        return sum([line.product_qty for line in tracking.move_ids])
+
+    def _get_measurement(self, tracking_id):
+        tracking = self.pool.get('stock.tracking').browse(self.cr, self.uid, tracking_id)
+        return r'*'.join([str(tracking.pack_h), str(tracking.pack_w), str(tracking.pack_l)] )
```
Comment (gurneyalex) [inline stock_pack_wizard/report/stock_tracking_list.py:61]: note: using str() may not give you the correct rounding. 
and you probably should use u'\u00d7' (multiplication sign, nicer than '*')

File: stock_pack_wizard/report/stock_trackling_list.mako [hunk 19]
```diff
@@ -0,0 +1,250 @@
...
+
+</head>
+<body>
+
+<% 
+sum_pack=0
+sum_qty=0
+sum_nw=0
+sum_gw=0
+sum_cbm=0
```
Comment (gurneyalex) [inline stock_pack_wizard/report/stock_trackling_list.mako:99]: why are you not using the values you compute in the context of the repport?

File: stock_pack_wizard/wizard/wizard_picking_tracking.py [hunk 20]
```diff
@@ -0,0 +1,362 @@
...
+        return default_ul_id
+
+    _columns = {
+        'name': fields.char('name', size=32, readonly=True),
+        'picking_id': fields.many2one('stock.picking', 'Picking'),
+        'ul_id': fields.many2one('product.ul', 'Pack Template', required=True),
+        'lines': fields.one2many(
+            'wizard.picking.tracking.line', 'wizard_id', 'Lines'),
+        'pack_address': fields.char('Address', size=128),
+        'pack_note': fields.char('Note', size=128),
```
Comment (gurneyalex) [inline stock_pack_wizard/wizard/wizard_picking_tracking.py:234]: same remark as above regarding field size.

File: stock_pack_wizard/wizard/wizard_picking_tracking.py [hunk 21]
```diff
@@ -0,0 +1,362 @@
...
+    }
+
+    def _action_check(self, wizard):
+        '''Do some checks:
+        - picking should not be in done state
+        - pack quantity should always smaller than delivery quantity'''
+        if wizard.picking_id.state == 'done':
+            raise orm.except_orm(
+                _('Warning!'),
+                _("Wizard split pack can not be used when state is done !"))
```
Comment (gurneyalex) [inline stock_pack_wizard/wizard/wizard_picking_tracking.py:253]: don't put exclamation marks.

**Comment by gurneyalex** [comment]
nice module. A few things need fixing, including the tests in travis.

**Comment by duanyp1991** [comment]
@gurneyalex  Thanks for your advise! I'll come back with the improvement soon.

**Comment by Cedric-Pigeon** [comment]
@duanyp1991 Do you plan to achieve this PR, otherwise we will close it in 3 days. Thanks

**Comment by pedrobaeza** [comment]
No answer, so I close it.

**Pull request closed without merging**